### PR TITLE
Add tree-sitter json grammar

### PR DESCRIPTION
**Pull request by Ben3eeE**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

Depends on https://github.com/tree-sitter/tree-sitter-json/pull/6

- [x] Test bundled themes
- [x] Test community syntax themes

### Benefits

Enhanced code folding and syntax highlighting.

### Possible Drawbacks

* Less links are styled as links because the regex is less advanced than textmates `language-hyperlink` regex

### Applicable Issues

No

### CODE REVIEW

**Comment by simurai** [comment]
@Ben3eeE Made a PR for "Fix key/value coloring in one-syntax" here: https://github.com/atom/atom/pull/18515

**Comment by Ben3eeE** [comment]
I think `#1` and `#2` sounds best. Maybe we can open issues on all the other themes that we don't submit a PR to.

**Comment by Ben3eeE** [comment]
@simurai Alternative solution: Scope `object` as `meta.structure.dictionary.json` and add `json` to the key. It gives the same structure as textmate and works 😅 

It does apply the `meta.structure.dictionary.json` scope multiple times for nested objects but I think that should be fine.

**Comment by simurai** [comment]
> Alternative solution

Ohhh.. awesome. Tested and yes, works without making changes to themes. 🙌 If this still makes sense, we should definitely take this route. 👍 

I'll close https://github.com/atom/atom/pull/18515 again to not keep "stale" PRs around.

File: grammars/tree-sitter-json.cson [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+scopes:
+  'value': 'source.json'
+
+  'object': 'meta.structure.dictionary.json'
+
+  'pair > string': [
+      {
+        match: '^"https?://',
+        scopes: 'markup.underline.link'
```
[the remark this answers is not in the document]
Comment (Ben3eeE) [reply grammars/tree-sitter-json.cson]: @50Wliu https://github.com/atom/atom/issues/18196#issuecomment-432741331

**Comment by caleb531** [comment]
Hi,

What is the status of this PR? I would love to see JSON files using tree-sitter in core Atom, since the JSON files I work with are often very large. Thanks in advance!

**Comment by rsese** [comment]
Hey @caleb531 :wave:  The PR is on the team's list to review but there's no specific timeframe for it with the work on their plate right now.  I can mention that you checked in on it though.

If you'd like, you can see what's being worked on in the biweekly [iteration plans](https://github.com/atom/atom/labels/iteration-plan).

**Comment by Ben3eeE** [comment]
I'll merge this in a few days based on https://github.com/atom/language-python/pull/288#issuecomment-481871525

Unless someone objects.

**Comment by caleb531** [comment]
@Ben3eeE It looks like array strings are completely un-tokenized. Could you please update the grammar to give them color? Numbers and booleans look fine.

**Strings in an object:**  
Screen Shot 2019-04-13 at 10 16 23 AM

**Strings in an array:**  
Screen Shot 2019-04-13 at 10 16 35 AM

**Syntax tree for array strings (for quick reference):**  
Screen Shot 2019-04-13 at 10 17 06 AM

**Comment by Ben3eeE** [comment]
> It looks like array strings are completely un-tokenized. Could you please update the grammar to give them color? Numbers and booleans look fine.

Thanks for pointing this out, you are correct 🙂

**Comment by caleb531** [comment]
@Ben3eeE Is it possible to highlight URL strings as `string.quoted.double` in addition to `markup.underline.link`? I ask this because if the theme does not have styling for `markup.underline.link`, then those URL strings currently show up as white (i.e. untokenized) rather than as strings.

For example, in my recent PR to the Monokai theme (https://github.com/kevinsawicki/monokai/pull/105), I had to explicitly add styling for `markup.underline.link` because those matched strings would otherwise not be tokenized as anything.

**Comment by Ben3eeE** [comment]
@caleb531 Thank you so much for testing this out 🙇 I intend to take a look at the PR you opened on the syntax theme to see if we can fix it here before merging.

If you want to check out more syntax themes that would be great 🙂 I'll take another look to see if there are any regressions.

**Comment by caleb531** [comment]
@Ben3eeE I've tested several of the featured, most popular, and default syntax themes. Please see the screenshots below, followed by my thoughts on where to proceed from here (TL;DR—I propose we remove `markup.underline.link` tokenization entirely and just highlight as strings):

**one-dark:**
one-dark

**one-light:**  
one-light

**atom-dark:**  
atom-dark

**base16-tomorrow-dark:**  
base16-tomorrow-dark

**atom-material-syntax:**  
atom-material-syntax

**seti-syntax:**  
seti-syntax

Considering the above, I ultimately recommend we remove tokenization for links entirely in this tree-sitter JSON grammar, for several reasons:

1. As shown above, the most popular themes are missing styles for `markup.underline.link`. I know this could be fixed with PRs, but by removing the tokenization and simply tokenizing as strings, we eliminate the need to do this, and we can get this grammar merged sooner (provided you make the changes to the key structure, as discussed above)
2. The quotes are highlighted as `markup.underline.link`, which is semantically incorrect—only the URL _between_ the quotes should be tokenized as a link, though I recognize that the tree-sitter-json syntax tree doesn't assign anything to the text between the quotes
3. The JSON format doesn't give any special meaning to URLs, so they shouldn't be colored differently than other string values
4. The first-mate JSON grammar doesn't tokenize links (the core `language-hyperlink` package handles any link highlighting you might see for first-mate JSON files); I would prefer to wait for the tree-sitter injection system, which will need to happen anyway for tree-sitter to eventually replace first-mate

Please let me know your thoughts on this—questions, comments or concerns are welcome.

**Comment by Arcanemagus** [comment]
> As shown above, the most popular themes are missing styles for markup.underline.link.

Many themes are missing _many_ tokens, what about this is causing a problem that would cause that tokenization to be removed?

> The quotes are highlighted as markup.underline.link, which is semantically incorrect—only the URL between the quotes should be tokenized as a link

👍

> The JSON format doesn't give any special meaning to URLs, so they shouldn't be colored differently than other string values

Now _this_ is a valid argument for removing URL highlighting as you propose in `1.`. However we should probably still tokenize them as...

> The first-mate JSON grammar doesn't tokenize links (the core language-hyperlink package handles any link highlighting you might see for first-mate JSON files); I would prefer to wait for the tree-sitter injection system, which will need to happen anyway for tree-sitter to eventually replace first-mate

This was added in https://github.com/atom/atom/pull/17551, supported since Atom v1.30.0 hit stable. As far as I understand it if no other grammar is in place to handle the injection it acts as if the injection wasn't there (meaning the links would show as regular strings), but once a grammar is installed that understands that injection point it will start working automatically.

**Comment by caleb531** [comment]
@Arcanemagus 

> Many themes are missing _many_ tokens, what about this is causing a problem that would cause that tokenization to be removed?

That's fair—we may still end up submitting PRs to these themes anyway to resolve those other tokenization issues—so removing the URL tokenization wouldn't necessarily reduce the effort.

> This was added in atom/atom#17551, supported since Atom v1.30.0 hit stable. As far as I understand it if no other grammar is in place to handle the injection it acts as if the injection wasn't there (meaning the links would show as regular strings), but once a grammar is installed that understands that injection point it will start working automatically.

Do you think if we implement a proper tree-sitter injection (via `atom.grammars.addInjectionPoint`), we will still be able to tokenize the URL in JSON _minus the quotes_? Or will that ultimately require modifying `tree-sitter-json` to add a syntax node for the text between the quotes of a string node?

I suppose to implement the injection, we would also need to add a tree-sitter grammar to language-hyperlink? (because, as I recall, a first-mate grammar can't be injected into a tree-sitter grammar)

**Comment by Arcanemagus** [comment]
> Or will that ultimately require modifying tree-sitter-json to add a syntax node for the text between the quotes of a string node?

I _think_ that the injection happens for an entire node, so we would need to recognize the URL _area_ here first from how I understand it, but I don't have personal experience with how it works exactly.

> I suppose to implement the injection, we would also need to add a tree-sitter grammar to language-hyperlink? (because, as I recall, a first-mate grammar can't be injected into a tree-sitter grammar)

Correct, that's a known limitation (feature? 😆) of this system that it has no way of embedding a TextMate grammar.

**Comment by Ben3eeE** [comment]
> The quotes are highlighted as markup.underline.link, which is semantically incorrect—only the URL between the quotes should be tokenized as a link, though I recognize that the tree-sitter-json syntax tree doesn't assign anything to the text between the quotes


Yeah they should be tokenized differently. tree-sitter does recognize them as an anonymous `"` node but we can't scope them because of a bug somewhere in https://github.com/atom/atom/blob/bc7ec2d3a83b2616713ba7e929e63552b0809c84/src/syntax-scope-map.js

I tried scoping:
`'"\""': 'punctuation.blabla'`

Which seems like it should work but it doesn't. The grammar fails to load with an unclosed quote error.

If we can fix this bug and scope anonymous `"` nodes they would be tokenized differently.

We have a similar issue in `language-c`:
https://github.com/atom/language-c/blob/a355a4d4a019d5684c40072b81f24c89e58dc56f/grammars/tree-sitter-c.cson#L142

**Comment by nathansobo** [comment]
After reviewing the code and the discussion, it seems to me that this is ready.

**Comment by nathansobo** [comment]
Thanks @Ben3eeE and sorry for the wait. You are awesome.

**Comment by nathansobo** [comment]
Ugh. Assumed the AppVeyor failure was spurious but was not. Something changed. @50Wliu does [this failure](https://ci.appveyor.com/project/Atom/language-json/builds/24463865/job/8sxx24moavw772l2) look familiar at all based on apm stuff you've dealt within the past?

**Comment by nathansobo** [comment]
@50Wliu nevermind the previous comment. I solved the issue in #72. 😓

**Comment by Ben3eeE** [comment]
@nathansobo Thank you 🙇 I have been using this linked locally since I opened it because I have been working with large json files lately. The performance benefit with this is huge 👍.

**Comment by 50Wliu** [comment]
It doesn't seem like the issue with URL strings was ever resolved? Here's what I'm seeing, which is pretty jarring:
json-url-strings

**Comment by nathansobo** [comment]
I looked at our treatment of URLs in the other grammar and it looked similar. What am I missing?

**Comment by 50Wliu** [comment]
Here's how Textmate grammars handle links (thanks to language-hyperlink):
textmate-links
And the corresponding scopes:
textmate-link-scopes

Whereas this grammar has the following scopes:
tree-sitter-json-link-scopes

Notably, `string.quoted.double` is missing from the scope structure, so I only get the underline, not the foreground color that I've assigned to strings.

Additionally, as mentioned above, the quotes themselves should not be part of the link.

**Comment by Ben3eeE** [comment]
So we could add `string.quoted.double` to https://github.com/atom/language-json/pull/68/files#diff-fce1e3e82e57cd20261fef4754b35797R56 to add the scope and match textmate?

Quotes being included is an issue with every grammar because we can't scope them and tree-sitter recognizes quotes as part of the string.

**Comment by 50Wliu** [comment]
@Ben3eeE pointed out that my previous comment could be really confusing, so here's a better attempt.
Regular JSON strings (tree-sitter):
regular-json-strings-tree-sitter

URL JSON strings (tree-sitter):
url-json-strings-tree-sitter

Regular JSON strings (textmate):
regular-json-strings-textmate

URL JSON strings (textmate):
url-json-strings-textmate

So the key difference is that there's no `string.quoted.double` scope for the URL JSON tree-sitter case.

**Comment by nathansobo** [comment]
Talked to @Ben3eeE and he plans to address this issue tomorrow. #19336 should give him the tools he needs to style the quote literals explicitly, and it seems like adding `string.quoted.double` to the scope in the URL-specific `match` rule should also help.

**Pull request merged**